eventhandling/getSliEvent.go

```diff
@@ -0,0 +1,275 @@
+		return nil, errors.New("could not create Kubernetes client")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		log.Println("could not create Kubernetes client")
+		return nil, errors.New("could not create Kubernetes client")
+	}
+
+	prometheusAPIURL, err := getPrometheusAPIURL(eventData.Project, kubeClient.CoreV1())
+	if err != nil {
+		return nil, err
+	}
+
+	keptnHandler, err := keptnv2.NewKeptn(&event, keptncommon.KeptnOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new Prometheus Handler
+	prometheusHandler := utils.NewPrometheusHandler(
+		prometheusAPIURL,
+		&eventData.EventData,
+		eventData.GetSLI.CustomFilters,
+	)
+
+	projectCustomQueries, err := getCustomQueries(keptnHandler, eventData.Project, eventData.Stage, eventData.Service)
+	if err != nil {
+		log.Println("retrieveMetrics: Failed to get custom queries for project " + eventData.Project)
+		log.Println(err.Error())
+		return nil, err
+	}
+
+	if projectCustomQueries != nil {
+		prometheusHandler.CustomQueries = projectCustomQueries
+	}
+
+	var sliResults []*keptnv2.SLIResult
+
+	for _, indicator := range eventData.GetSLI.Indicators {
+		log.Println("retrieveMetrics: Fetching indicator: " + indicator)
+		sliValue, err := prometheusHandler.GetSLIValue(indicator, eventData.GetSLI.Start, eventData.GetSLI.End)
+		if err != nil {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   0,
+				Success: false,
+				Message: err.Error(),
+			})
+		} else if math.IsNaN(sliValue) {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   0,
+				Success: false,
+				Message: "SLI value is NaN",
+			})
+		} else {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   sliValue,
+				Success: true,
+			})
+		}
+	}
+	return sliResults, nil
+}
+
+func getCustomQueries(keptnHandler *keptnv2.Keptn, project string, stage string, service string) (map[string]string, error) {
+	log.Println("Checking for custom SLI queries")
+
+	customQueries, err := keptnHandler.GetSLIConfiguration(project, stage, service, utils.SliResourceURI)
+	if err != nil {
+		return nil, err
+	}
+
+	return customQueries, nil
+}
+
+// getPrometheusAPIURL fetches the prometheus API URL for the provided project (e.g., from Kubernetes configmap)
+func getPrometheusAPIURL(project string, kubeClient v1.CoreV1Interface) (string, error) {
+	log.Println("Checking if external prometheus instance has been defined for project " + project)
+	secret, err := kubeClient.Secrets(env.PodNamespace).Get("prometheus-credentials-"+project, metav1.GetOptions{})
+
+	// return cluster-internal prometheus URL if no secret has been found
+	if err != nil {
+		log.Println("could not retrieve or read secret: " + err.Error())
+		log.Println("No external prometheus instance defined for project " + project + ". Using default: " + env.PrometheusEndpoint)
```

Comment:
shouldn't these 2 logs be the same? The cause is the error while reading the secret so a single log saying something like:

`log.Printf("could not retrieve or read prometheus credential secret for project %s: %v. Using default external prometheus endpoint: %s", project, err, env.PrometheusEndpoint)`

could be less confusing...

Reply:
Good point, done